Sources/Deployer/DeploymentDestination.swift

```diff
@@ -6,14 +6,16 @@ public struct DeploymentDestination: Codable, CustomStringConvertible, Hashable
     public let host: String
     public let port: Int32
     public let username: String
-    public let password: String
+    public let password: String?
+    public let key: String?
```

Comment:
There are only two valid ways to perform auth here: username+password or key. But three distinct fields define much more combinations. What will happen if I define username+key? username only?

I suggest to introduce an `enum`, e.g.

```swift
public enum DeploymentDestinationAuthenticationType {
    case plain(username: String, password: String)
    case key(String)
}
```

Also, even after reading all changes in this PR, it is not obvious whether `key` should contain an actual key value or it should contain a path to a file with a key. Could you please make this more clear, e.g. 
```swift
case key(path: String)
``` 
or
```swift
case key(keyString: String)
```
or something like that.

You probably will need to write a code for encoding and decoding this to/from JSON, here is [an example](https://github.com/avito-tech/Emcee/blob/master/Sources/AutomaticTermination/AutomaticTerminationPolicy.swift)